File: BOJ/11724_1.cpp
// 연결 요소의 갯수
// https://www.acmicpc.net/problem/11724
#include <iostream>
#include <vector>
using namespace std;
const int MAX = 1000 + 1;
int N, M;
vector<int> graph[MAX];
bool visited[MAX];
void dfs(int current)
{
    visited[current] = true;
    for (int i = 0; i < graph[current].size(); i++)
    {
        int next = graph[current][i];
        if (!visited[next])
            dfs(next);
    }
}
void init()
{
    cin >> N >> M;
    int u, v;
    for (int i = 0; i < M; i++)
    {
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
}
int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    init();
    int answer = 0;
    for (int i = 1; i <= N; i++)
    {
        if (!visited[i])
        {
            dfs(i);
            answer++;
        }
    }
    cout << answer;
    return 0;
}

File: BOJ/11724_1.py
'''
    연결 요소의 갯수
    https://www.acmicpc.net/problem/11724
'''
import sys
input = sys.stdin.readline
sys.setrecursionlimit(1000000)
N, M = map(int, input().split())
visited = [0] * (N + 1)
graph = [[] * (M) for _ in range(N+1)]
for _ in range(M):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)


def dfs(current):
    visited[current] = True
    for i in range(len(graph[current])):
        next = graph[current][i]
        if visited[next] == False:
            dfs(next)


answer = 0
for i in range(1, N + 1):
    if visited[i] == False:
        dfs(i)
        answer += 1
print(answer)
